Hyperledger Fabric 2.x with new chaincode lifecyle

Kelvin-M opened the pull request:
Migration to HLF 2.X with new chaincode lifecycle, etcdraft and couchdb.
- Update HLF fabric alpine images (orderer, peer and tools) to 2.2.0 and HL fabric CA to 1.4.9
- Use modified HLF fabric-peer image to add external chaincode configuration in core.yaml and binaries
- Create two hlf-k8s images (2.2.0 for tool + 1.4.8 for CA enrollment operator)
- Update configtx for genesis and application channel (capabilities, policies are mandatory) and core.yaml to last version 
- Migrate from Solo to Raft : configtx to handle raft
- Use couchdb instead of goleveldb with hlf-couchdb chart.
- Use new chaincode lifecycle : external chaincode packaging, update chaincode operator and app channel operator

Companion PR : 
https://github.com/SubstraFoundation/substra-backend/pull/343
https://github.com/SubstraFoundation/substra-chaincode/pull/127
https://github.com/SubstraFoundation/substra-tests/pull/157

Kelvin-M commented on examples/dev-secrets.sh:
```diff
@@ -34,3 +34,14 @@ ${KUBECTL} "$OP" -f "${DIR}/secrets/secrets-orderer-genesis.yaml"
+
+
+
+if [ "$OP" = "delete" ]; then
+    ${KUBECTL} get secrets -A | grep chaincode | awk '{print $1,$2}' | xargs -n2 ${KUBECTL} delete secret -n
+fi
```
Improve that to make a hook or something else !

Kelvin-M commented on examples/3-orgs-policy-majority/skaffold.yaml:
```diff
@@ -14,7 +14,7 @@
 # ABOUT
 #
-#   This is An hlf-k8s deployment with 3 organizations using the default
+#   This is An hlf-k8s deployment with 4 organizations using the default
```
3 here

Kelvin-M commented on examples/3-orgs-policy-majority/values/orderer.yaml:
```diff
@@ -26,13 +26,13 @@ hlf-ord:
     - { org: MyOrg1, mspid: MyOrg1MSP, configUrl: network-org-1-peer-1-hlf-k8s-config-operator.org-1/config/configOrg.json }
     - { org: MyOrg2, mspid: MyOrg2MSP, configUrl: network-org-2-peer-1-hlf-k8s-config-operator.org-2/config/configOrg.json }
     - { org: MyOrg3, mspid: MyOrg3MSP, configUrl: network-org-3-peer-1-hlf-k8s-config-operator.org-3/config/configOrg.json }
+    - { org: MyOrg4, mspid: MyOrg4MSP, configUrl: network-org-4-peer-1-hlf-k8s-config-operator.org-4/config/configOrg.json }
```
Remove that

AurelienGasser reviewed (commented):
Partial review

Kelvin-M commented on charts/hlf-k8s/templates/deployment-chaincode-operator.yaml:
```diff
@@ -40,48 +41,103 @@ spec:
+                  \"root_cert\": \"$(awk 'NF {sub(/\r/, ""); printf "%s\\n",$0;}' /var/hyperledger/tls/client/cert/cacert.pem)\"
+                  }" > connection.json
+
+                  cat connection.json
```
Yes, it's simpler for administrator to check if the connection.json file is well done. I can remove it if not necessary !

Kelvin-M commented on charts/hlf-k8s/templates/deployment-chaincode-operator.yaml:
```diff
@@ -40,48 +41,103 @@ spec:
+              printf "[DEBUG] Testing the connection with the Peer ({{ index $.Values "hlf-peer" "host" }}:{{ index $.Values "hlf-peer" "port" }})\n"
+              until $(nc -z {{ index $.Values "hlf-peer" "host" }} {{ index $.Values "hlf-peer" "port" }}); do
+                printf "[DEBUG] Peer ({{ index $.Values "hlf-peer" "host" }}:{{ index $.Values "hlf-peer" "port" }}) is not reacheable, retry in 5s\n"
+                sleep 5
```
Sleep 5 seems good as waiting only for 1 second is maybe too short

AurelienGasser replied:
Why too short?

Kelvin-M replied:
I don't know what is the good time to sleep. But it will take to much time to do a benchmark to know what is the good value for the sleep, and this value of 5 is used everywhere we wait for a service like a peer or an orderer :)

AurelienGasser replied:
I measured current v.s. "without sleep" and here are the results:

```
# Actuel: avec `sleep X` (X > 0) 
real    0m30.736s
real    0m34.121s
real    0m32.005s
# Test: avec `sleep 0.1`
real    0m20.118s
real    0m28.689s
real    0m30.416s
real    0m25.137s 
```

There's a difference, but it's overall minor.

AurelienGasser replied:
I measured the deployment stabilization time + the time for the chaincode to be ready in the "current" v.s. "without sleep" scenarios. 

Here are the results:

```
# Current: with `sleep X` (X > 0) 
Deployments stabilized in 26.899945242s
real    0m3.097s
Deployments stabilized in 26.560970816s
real    0m3.188s
Deployments stabilized in 28.159576478s
real    0m3.145s
Deployments stabilized in 24.25722498s
real    0m3.176s
Deployments stabilized in 25.591797322s
real    0m3.230s

# Test: with `sleep 0.1` (I replaced all the sleep X with sleep 0.1 in the code)
Deployments stabilized in 2m23.142196021s
real    0m5.069s
Deployments stabilized in 15.300134681s
real    0m4.595s
Deployments stabilized in 9.724714045s
real    0m4.274s
Deployments stabilized in 18.236987181s
real    0m4.642s
Deployments stabilized in 8.295457046s
real    0m4.399s
```

Test code: 

```
for i in `seq 4`; do 
    ./examples/dev-secrets.sh create
    sk run
    { time while true; do ./examples/test-dev-network.sh && break;done } 2>&1 | tee -a with-sleep.txt
    sk delete
done
````

Kelvin-M replied:
I see `2m23.142196021s`, looks big no ?
What is the conclusion of your test ?

AurelienGasser replied:
Yeah that's an outlier, I'm not sure what happened there.

> What is the conclusion of your test ?

Excluding the outlier, the deployment is 10-15 sec faster without the sleeps. It's just a data point to have an idea of how `sleep`s affect deployment time.

Kelvin-M replied:
Yes, I think we can do this optimisation in another PR :)

Kelvin-M commented on charts/hlf-k8s/templates/rbac.yaml:
```diff
@@ -42,9 +42,7 @@ rules:
     - watch
     - list
     - create
-    {{- if .Values.hooks.deleteSecrets.enabled }}
```
Because there is a job hook to remove chaincode ccid secrets now :)

AurelienGasser replied:
Shouldn't this condition be kept but modified instead (so that if any of the hooks is enabled we add the "delete secrets" permission).

My understanding is that now the "delete" permission is granted even when it's not needed (i.e. when both hooks are disabled)

Kelvin-M commented on charts/hlf-k8s/templates/deployment-appchannel-operator.yaml:
```diff
@@ -125,18 +126,29 @@ spec:
+                    printf "[DEBUG] Add the organization ($org) for endorsement \n"
+                    PEER_HOST=$(jq -r .values.AnchorPeers.value.anchor_peers[0].host configOrg-$org.json)
+                    PEER_PORT=$(jq -r .values.AnchorPeers.value.anchor_peers[0].port configOrg-$org.json)
+                    grep -qxF -e "--peerAddresses ${PEER_HOST}:${PEER_PORT} --tlsRootCertFiles tlsRootCert-$org.crt" endorsement.config || echo "--peerAddresses ${PEER_HOST}:${PEER_PORT} --tlsRootCertFiles tlsRootCert-$org.crt" >> endorsement.config
```
In grep '-' is interpreted as an option, 

> -e  : The following argument is taken as a pattern. This option makes it easy to specify patterns that might confuse argument parsing, such as -n.

AurelienGasser replied:
:+1:

Kelvin-M commented on charts/hlf-k8s/templates/deployment-appchannel-operator.yaml:
```diff
@@ -210,20 +222,64 @@ spec:
+
+                peer lifecycle chaincode commit --signature-policy "{{ .chaincodePolicy}}" --channelID {{ $value.channelName }} --name {{ .chaincodeName }} --version {{ .chaincodeVersion }} --sequence 1 --init-required -o {{ index $.Values "hlf-ord" "host" }}:{{ index $.Values "hlf-ord" "port" }} --tls --clientauth --cafile /var/hyperledger/tls/ord/cert/cacert.pem --keyfile /var/hyperledger/tls/c ...
+
+                cat chaincode.{{ .chaincodeName }}.commit
```
Yes, in case of error during commit, it's better to see why in logs :)

Kelvin-M commented on charts/hlf-k8s/templates/deployment-appchannel-operator.yaml:
```diff
@@ -210,20 +222,64 @@ spec:
+
+                peer lifecycle chaincode checkcommitreadiness --signature-policy "{{ .chaincodePolicy}}" --channelID {{ $value.channelName }} --name {{ .chaincodeName }} --version {{ .chaincodeVersion }} --sequence 1 --init-required -o {{ index $.Values "hlf-ord" "host" }}:{{ index $.Values "hlf-ord" "port" }} --tls --clientauth --cafile /var/hyperledger/tls/ord/cert/cacert.pem --keyfile /var/hyp ...
+
+                sleep 5
```
Yes, I will update the code 👍

Kelvin-M commented on charts/hlf-k8s/templates/configmap-fabric.yaml:
```diff
@@ -29,54 +29,135 @@ data:
+          path: /builders
+          propagateEnvironment:
+             - GOPROXY
+             - ENVVAR_NAME_TO_PROPAGATE_FROM_PEER
```
It may be useless in our case. I think I add it when I incrementally switch from dockered chaincode to k8sed one :D

Kelvin-M replied:
Removed

AlexandrePicosson commented on charts/hlf-k8s/values.yaml:
```diff
@@ -12,10 +12,16 @@
+    tag: latest
+    pullPolicy: IfNotPresent
+
+  hlfk8sca:
```
would `hlfk8sCa` be more readable and consistent here ? As we use lowerCamelCase everywhere else.

Kelvin-M replied:
Yes :)

AlexandrePicosson commented on charts/hlf-k8s/templates/deployment-enrollement-operator.yaml:
```diff
@@ -37,7 +37,7 @@ spec:
       containers:
         - name: fabric-tools
-          image: {{ .Values.image.repository }}:{{ .Values.image.tag }}
+          image: {{ .Values.images.hlfk8sca.repository }}:{{ .Values.images.hlfk8sca.tag }}
```
I find this line confusing, reading the code quickly we could think that we are using a CA image for the enrollment.
We could use something like `hlfk8sCaClient`.

Kelvin-M replied:
I agree for `hlfk8sCaEnrollment` as it is only used for CA enrollement. Nevertheless `hlfk8s` image is used for multiple other deployment (for instance toolbox)

AlexandrePicosson replied:
Yes, there is no issue with `hlfk8s`, reading the code we can clearly see it is a multi-purpose image with tooling.

Just for information, why did you choose to use a different image here and not add more tools to the hlfk8s image ?

Kelvin-M replied:
Because, the two images are not based on the same version (fabric 2.2 and 1.4.9)
CA does not have a version increase with HLF 2.x

AlexandrePicosson commented on charts/hlf-k8s/README.md:
```diff
@@ -120,15 +126,19 @@ Install a chaincode on a peer using the following values.
-    # Chaincode source code archive URL
-    src: https://github.com/SubstraFoundation/substra-chaincode/archive/0.0.2.tar.gz
+appChannels:
+    chaincodes:
```
```suggestion
  - channelName: mychannel
    chaincodes:
```
Showing the fact that a chaincode lives inside a channel could be a good idea, what do you think ?

Kelvin-M replied:
Yes !

inalgnu commented on charts/hlf-k8s/values.yaml:
```diff
@@ -137,7 +180,12 @@ hlf-ca:
+    enabled:
+      false
   peer:
+    databaseType: CouchDB # goleveldb
```
this comment still relevent?

Kelvin-M replied:
Nope !

AurelienGasser commented on images/fabric-peer/builders/bin/build:
```diff
@@ -0,0 +1,34 @@
+if [ -d "$CHAINCODE_SOURCE_DIR/metadata" ]; then
+    cp -a $CHAINCODE_SOURCE_DIR/metadata $BUILD_OUTPUT_DIR/metadata
+fi
+
+exit 0
```
Nit: that's unnecessary

AurelienGasser commented on examples/test-dev-network.sh:
```diff
@@ -29,3 +29,20 @@ for i in `seq $NUM_ORGS`; do
         -c '{\"Args\":[\"queryTraintuples\"]}'"
     echo '-----------'
 done
+
```
Is the chaincode already approved/commited at the point in time when the chaincode is instantiated (i.e. the chaincode pod is running and Ready?)

If so, maybe we don't need this file at all anymore? In other words, the chaincode pods being up-and-running guarantees that the chaincode is invokable.

Kelvin-M replied:
Chaincode is running and ready by itself and only need package chaincode CCID.
Approval and commit steps are for the channel to allow transaction from that chaincode.

AurelienGasser commented on charts/hlf-k8s/requirements.yaml:
```diff
@@ -14,18 +14,22 @@
     condition: hlf-peer.enabled
   - name: nginx-ingress
-    repository: https://kubernetes-charts.storage.googleapis.com/
+    repository: https://charts.helm.sh/stable
```
I think we need a PR in substra-tests to add `https://charts.helm.sh/stable` and `https://owkin.github.io/charts/`

Kelvin-M replied:
Yes :) 
https://github.com/SubstraFoundation/substra-tests/pull/157

AurelienGasser reviewed (changes requested):
Awesome work!

I added a few comments inline. 

I think the two other things that are missing are:

**1. Support for custom local chaincodes**

This could be in the form of documentation, or ideally be automated. The use-case to support is "a developer edits `substra-chaincode` code locally and wants to test it in `hlf-k8s`."

**2. Support for end-to-end tests**

This problem might overlap with point 1. In order to test end-to-end flows automatically, we need the ability to specify a custom `substra-chaincode` branch in our end-to-end tests.

I'm happy to take some time with you to discuss possible options to address these issues!

Kelvin-M commented on examples/dev-secrets.sh:
```diff
@@ -34,3 +34,8 @@ ${KUBECTL} "$OP" -f "${DIR}/secrets/secrets-orderer-genesis.yaml"
+${KUBECTL} "$OP" -f "${DIR}/secrets/secrets-orderer-genesis.yaml"
+${KUBECTL} "$OP" -f "${DIR}/secrets/secrets-orderer.yaml"
+${KUBECTL} "$OP" -f "${DIR}/secrets/secrets-org-1.yaml"
+${KUBECTL} "$OP" -f "${DIR}/secrets/secrets-org-2.yaml"
```
Nope, surely some remaining test code

Kelvin-M commented on images/fabric-peer/builders/bin/detect:
```diff
@@ -0,0 +1,25 @@
+
+set -euo pipefail
+
+# use jq to extract the chaincode type from metadata.json and exit with
+# success if the chaincode type is golang
```
I didn't use this website, but this one : https://medium.com/swlh/how-to-implement-hyperledger-fabric-external-chaincodes-within-a-kubernetes-cluster-fd01d7544523 :D

Kelvin-M commented:
> Awesome work!
> 
> I added a few comments inline.
> 
> I think the two other things that are missing are:
> 
> **1. Support for custom local chaincodes**
> 
> This could be in the form of documentation, or ideally be automated. The use-case to support is "a developer edits `substra-chaincode` code locally and wants to test it in `hlf-k8s`."
> 
> **2. Support for end-to-end tests**
> 
> This problem might overlap with point 1. In order to test end-to-end flows automatically, we need the ability to specify a custom `substra-chaincode` branch in our end-to-end tests.
> 
> I'm happy to take some time with you to discuss possible options to address these issues!

Point 2 is addressed here : https://github.com/SubstraFoundation/substra-tests/pull/157
Point 1 is now addressed in the charts readme

AurelienGasser commented on charts/hlf-k8s/README.md:
````diff
@@ -120,14 +126,55 @@ Install a chaincode on a peer using the following values.
+git clone git@github.com:SubstraFoundation/substra-chaincode.git
+```
+
+Do your modifications in substra-chaincode
+
+Then you have to build your own substra-chaincode image
+
+Do a `eval $(minikube -p minikube docker-env)` for Minikube user to use docker from it. For Docker-desktop user, you don't have to do this !
+
+Then build substra-chaincode image
+
+```bash
+docker build -t substrafoundation/substra-chaincode:my-tag ./
+```
+
+You should modify values from your deployement to use your image for the chaincode part
````
The code suggestion feature seems to have a bug because there's a code block within the suggestion. But you get the idea!

AurelienGasser reviewed (changes requested):
Nice additions!

There is a section `### Install a custom chaincode` in the root README that needs to be deleted.

ClementGautier commented on charts/hlf-k8s/README.md:
```diff
@@ -25,30 +25,34 @@ The following table lists the configurable parameters of the hlf-k8s chart and d
+| `appChannels[].channelPolicies` | This value overrides the default HLF channel policy. | (defined in values.yaml) |
+| `appChannels[].appPolicies` | This value overrides the default HLF application policy. | (defined in values.yaml) |
+| `appChannels[].chaincodes` | The chaincodes to install on the Peer. See [Install a chaincode](#install-a-chaincode). | `[]` |
+| `appChannels[].chaincodes[].chaincodeName` | The name of the chaincode | (undefined) |
+| `appChannels[].chaincodes[].chaincodeVersion` | The chaincode version | (undefined) |
+| `appChannels[].chaincodes[].chaincodeAddress` | The URL to the chaincode service | (undefined) |
+| `appChannels[].chaincodes[].chaincodePort` | The port to the chaincode service | (undefined) |
+| `appChannels[].chaincodes[].chaincodeAddress` | The URL to a chaincode service | (undefined) |
+| `appChannels[].chaincodes[].chaincodePolicy` | The chaincode policy | (undefined) |
```
I would just name them name/version/address/etc. without the "chaincode" prefix as you did for the image.

ClementGautier commented on charts/hlf-k8s/README.md:
```diff
@@ -74,16 +78,17 @@ The following table lists the configurable parameters of the hlf-k8s chart and d
-| `image.repository` | `hlf-k8s` image repository | `substrafoundation/hlf-k8s` |
-| `image.tag` | `hlf-k8s` image tag | `latest` |
-| `image.pullPolicy` | Image pull policy | `IfNotPresent` |
+| `images.hlfk8s.repository` | `hlf-k8s` image repository | `substrafoundation/hlf-k8s` |
+| `images.hlfk8s.tag` | `hlf-k8s` image tag | `latest` |
+| `images.hlfk8s.pullPolicy` | Image pull policy | `IfNotPresent` |
+| `images.hlfk8sCaEnrollment.repository` | `hlf-k8s-ca` image repository | `substrafoundation/hlf-k8s-ca` |
+| `images.hlfk8sCaEnrollment.tag` | `hlf-k8s-ca` image tag | `latest` |
+| `images.hlfk8sCaEnrollment.pullPolicy` | Image pull policy | `IfNotPresent` |
```
I would name that fabric-tools.image and fabric-ca-tools.image (at least put the "image" as a subpath of the service)

ClementGautier commented on charts/hlf-k8s/README.md:
```diff
@@ -101,7 +106,8 @@ The following table lists the configurable parameters of the hlf-k8s chart and d
 | `privateCa.configMap.fileName` | The CA certificate filename within the ConfigMap | `private-ca.crt` |
-| `hooks.deleteSecrets.enabled` | If true, the secrets created by the chart will be automatically deleted when the chart is uninstalled | `true` |
+| `hooks.deleteSecrets.enabled` | If true, the HLF crypto materials secrets created by the chart will be automatically deleted when the chart is uninstalled | `true` |
+| `hooks.deleteCCIDSecrets.enabled` | If true, the chaincode ccid secrets created by the chart will be automatically deleted when the chart is uninstalled | `true` |
```
why create an additionnal flag? I would have used deleteSecrets for simplicity here

Kelvin-M replied:
Because, you may want to keep hlf-secrets (crypto materials) but not chaincode CCID which is directly link to the installation of the chaincode

ClementGautier commented on charts/hlf-k8s/README.md:
```diff
@@ -120,14 +126,55 @@ Install a chaincode on a peer using the following values.
+appChannels:
+  - channelName: mychannel
+    chaincodes:
+      - chaincodeName: mycc
+        chaincodeVersion: "1.0"
+        chaincodeAddress: "chaincode-org-0-substra-chaincode-chaincode.org-0"
+        chaincodePort: "7052"
+        chaincodePolicy: "OR('Org1MSP.member','Org2MSP.member')"
+        image:
```
same, I would simply remove the chaincode prefix here to keep the keys clean

ClementGautier commented on charts/hlf-k8s/templates/deployment-appchannel-operator.yaml:
```diff
@@ -210,20 +222,67 @@ spec:
+                # Chaincode commit
+                printf "[DEBUG] Wait chaincode ready {{ .chaincodeName }} {{ .chaincodeVersion }} on channel {{ $value.channelName }}\n"
+
+                peer lifecycle chaincode checkcommitreadiness --signature-policy "{{ .chaincodePolicy}}" --channelID {{ $value.channelName }} --name {{ .chaincodeName }} --version {{ .chaincodeVersion }} --sequence 1 --init-required -o {{ index $.Values "hlf-ord" "host" }}:{{ index $.Values "hlf-ord" "port" }} --tls --clientauth --cafile /var/hyperledger/tls/ord/cert/cacert.pem --keyfile /var/hyp ...
```
For maintenance and readability I would rely on environment variables for all those options: https://hyperledger-fabric.readthedocs.io/en/release-2.2/enable_tls.html#configuring-tls-for-peers-nodes (this is for example the tls variables to set)

Kelvin-M replied:
Against an orderer, we cannot use environment variales : https://hyperledger-fabric.readthedocs.io/en/release-2.2/enable_tls.html#configuring-tls-for-the-peer-cli

```
When running a command that connects to orderer service, like peer channel <create|update|fetch> 
or peer chaincode <invoke>, following command line arguments must also be specified if TLS is enabled 
on the orderer:

–tls
–cafile <fully qualified path of the file that contains cert chain of the orderer CA>
If TLS client authentication is enabled on the orderer, the following arguments must be specified as well:

–clientauth
–keyfile <fully qualified path of the file that contains the client private key>
–certfile <fully qualified path of the file that contains the client certificate>
```

ClementGautier reviewed (changes requested):
Please could you add a UPGRADE.md file that sum up the major things to change when it come to migrate to this v4 ? For example the fact that you now need to put the chaincode fqdn in the csrHost of the enrollement is not obvious and I would have missed it if I'd not read that PR :D 

Great job for the PR, it's a ton of work!

Kelvin-M commented:
@AlexandrePicosson We will need to update dockerhub builds for hlf-k8s images :) !

ClementGautier commented on images/fabric-tools-ca/Dockerfile:
```diff
@@ -1,19 +1,20 @@
-FROM hyperledger/fabric-tools:1.4.6
+FROM hyperledger/fabric-tools:1.4.9
```
Why don't you start from https://hub.docker.com/r/hyperledger/fabric-ca-tools here ? Also if you could rename fabric-tools-ca to fabric-ca-tools that would be awsome :P (sorry for not catching it on the first review)

Kelvin-M replied:
Because it wasn't updated since two years.
Yes I can do the rename :)

ClementGautier replied:
Good point! :P

AurelienGasser reviewed (approved):
Nice fixes :)
